### PR TITLE
Document-fetcher using utils functions

### PR DESCRIPTION
**Pull request by prateekbh**
- Implements  common functions (`maybeIntercept_`, `fetchAmpCors_`, `fromStructuredCloneable_`, `toStructuredCloneable_`) as utility functions.
- `xhr-impl` and `document-fetcher` use these utility functions.
- No one uses document-fetcher for now(this is just to keep the files to review count as low as possible).
- Fixes Task #1 in #16707 

Next PR will shift all uses of `fetchDocument` to `document-fetcher`.

### CODE REVIEW

**Comment by prateekbh** [comment]
@jridgewell i've split it xhr class into utility functions, and everything seems to work, please have a look

File: src/utils/xhr-utils.js [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+ *
+ * @param {string} input The URL of the XHR to convert to structured
+ *     cloneable.
+ * @param {!FetchInitDef} init The options of the XHR to convert to structured
+ *     cloneable.
+ * @return {{input: string, init: !FetchInitDef}} The serialized structurally-
+ *     cloneable request.
+ * @private
+ */
+export function toStructuredCloneable(input, init) {
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: @jridgewell this function should be the exact copy of the previous code

File: src/utils/xhr-utils.js [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+ * `responseXML` as a `Document` type.
+ *
+ * @param {JsonObject|string|undefined} response The structurally-cloneable
+ *     response to convert back to a regular Response.
+ * @param {string|undefined} responseType The original response type used to
+ *     initiate the XHR.
+ * @return {!FetchResponse|!Response} The deserialized regular response.
+ * @private
+ */
+export function fromStructuredCloneable(response, responseType) {
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: @jridgewell this should be the exact copy as well, except a small change highlighted below

File: src/utils/xhr-utils.js [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+  user().assert(isObject(response), 'Object expected: %s', response);
+
+  const isDocumentType = responseType == 'document';
+  if (typeof Response === 'function' && !isDocumentType) {
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: this and response below used to be [win.Response](https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L309)

File: src/utils/xhr-utils.js [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+ * @param {?../service/ampdoc-impl.AmpDoc} ampdocSingle
+ * @param {string} input The URL of the XHR which may get intercepted.
+ * @param {!FetchInitDef} init The options of the XHR which may get
+ *     intercepted.
+ * @return {!Promise<!FetchResponse|!Response|undefined>}
+ *     A response returned by the interceptor if XHR is intercepted or
+ *     `Promise<undefined>` otherwise.
+ * @private
+ */
+export function getViewerInterceptResponse(win, ampdocSingle, input, init) {
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: @jridgewell this is just a renamed function of [maybeInterceptViewer](https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L184)

with removal of `this....` variables

File: src/utils/xhr-utils.js [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+}
+
+/**
+ * Setsup URL based on ampCors
+ * @param {!Window} win
+ * @param {string} input
+ * @param {!FetchInitDef} init The options of the XHR which may get
+ * intercepted.
+ */
+export function setupInput(win, input, init) {
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: this is an atomic function to modify input, which was here: https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L375-L377

File: src/utils/xhr-utils.js [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+}
+
+/**
+ * Sets up and normalizes the FetchInitDef
+ *
+ * @param {?FetchInitDef=} opt_init Fetch options object.
+ * @param {string=} opt_accept The HTTP Accept header value.
+ * @return {!FetchInitDef}
+ */
+export function setupInit(opt_init, opt_accept) {
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: This piece needs review.
Its a mix of logic operating on init here: https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L139-L142

and in original [setupInit](https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L562-L570) function

File: src/utils/xhr-utils.js [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+
+/**
+ *
+ * Sets up AMPSpecific CORS headers.
+ * @param {!Window} win
+ * @param {string} input
+ * @param {?FetchInitDef=} init
+ * @return {!FetchInitDef}
+ */
+export function setupAMPCors(win, input, init) {
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: Needs review
a function forked from here: https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L375-L396

File: src/utils/xhr-utils.js [hunk 8]
```diff
@@ -0,0 +1,624 @@
...
+  return init;
+}
+
+/**
+ * Normalized method name by uppercasing.
+ * @param {string|undefined} method
+ * @return {string}
+ * @private
+ */
+function normalizeMethod_(method) {
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: exact copy

File: src/utils/xhr-utils.js [hunk 9]
```diff
@@ -0,0 +1,624 @@
...
+}
+
+/**
+ * Verifies if response has the correct headers
+ * @param {!Window} win
+ * @param {!FetchResponse} response
+ * @param {!../utils/xhr-utils.FetchInitDef=} init
+ * @return {!FetchResponse}
+ */
+export function verifyAmpCORSHeaders(win, response, init) {
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: a post processing function forked from here: https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L398-L414

File: src/utils/xhr-utils.js [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+  } else if (init.requireAmpResponseSourceOrigin) {
+    // If the `AMP-Access-Control-Allow-Source-Origin` header is not
+    // returned but required, return error.
+    user().assert(false, 'Response must contain the' +
+        ` ${ALLOW_SOURCE_ORIGIN_HEADER} header`);
+  }
+  return response;
+}
+
+// TODO(prateekbh): move everything below this line into the polyfill
```
Comment (prateekbh) [inline src/utils/xhr-utils.js]: everything below this line should be exactly same

File: src/utils/xhr-utils.js [hunk 11]
```diff
@@ -0,0 +1,626 @@
...
+ * @param {!FetchInitDef} init The options of the XHR to convert to structured
+ *     cloneable.
+ * @return {{input: string, init: !FetchInitDef}} The serialized structurally-
+ *     cloneable request.
+ * @private
+ */
+export function toStructuredCloneable(input, init) {
+  const newInit = Object.assign({}, init);
+  if (isFormDataWrapper(init.body)) {
+    newInit.headers = newInit.headers || {};
```
Comment (jridgewell) [inline src/utils/xhr-utils.js]: Nit: This should be [guaranteed](https://github.com/ampproject/amphtml/pull/17206/files#diff-6870dab961227bda821a287600fb7158R281).

File: src/utils/xhr-utils.js [hunk 12]
```diff
@@ -0,0 +1,626 @@
...
+
+    if (init.credentials == 'include') {
+      xhr.withCredentials = true;
+    }
+
+    if (init.responseType in allowedFetchTypes_) {
+      xhr.responseType = init.responseType;
+    }
+
+    if (init.headers) {
```
Comment (jridgewell) [inline src/utils/xhr-utils.js]: Nit: This should be [guaranteed](https://github.com/ampproject/amphtml/pull/17206/files#diff-6870dab961227bda821a287600fb7158R281).

File: src/document-fetcher.js [hunk 13]
```diff
@@ -0,0 +1,145 @@
+/**
```
Comment (jridgewell) [inline src/document-fetcher.js]: Can we drop this file until the document fetch PR?

File: src/utils/xhr-utils.js (hunk 2 above)
Comment (prateekbh) [reply src/utils/xhr-utils.js]: not resolved.

File: src/utils/xhr-utils.js (hunk 1 above)
Comment (prateekbh) [reply src/utils/xhr-utils.js]: @jridgewell this function should be the exact copy of the previous code

File: src/utils/xhr-utils.js (hunk 3 above)
Comment (prateekbh) [reply src/utils/xhr-utils.js]: this and response below used to be win.Response

File: src/utils/xhr-utils.js (hunk 4 above)
Comment (prateekbh) [reply src/utils/xhr-utils.js]: @jridgewell this is just a renamed function of maybeInterceptViewer

with removal of this.... variables

File: src/utils/xhr-utils.js (hunk 5 above)
Comment (prateekbh) [reply src/utils/xhr-utils.js]: this is an atomic function to modify input, which was here: https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L375-L377

File: src/utils/xhr-utils.js (hunk 8 above)
Comment (prateekbh) [reply src/utils/xhr-utils.js]: exact copy

File: src/utils/xhr-utils.js [hunk 14]
```diff
@@ -0,0 +1,626 @@
...
+  }
+  if (init.requireAmpResponseSourceOrigin === undefined) {
+    init.requireAmpResponseSourceOrigin = true;
+  }
+  // For some same origin requests, add AMP-Same-Origin: true header to allow
+  // publishers to validate that this request came from their own origin.
+  const currentOrigin = getWinOrigin(win);
+  const targetOrigin = parseUrlDeprecated(input).origin;
+  if (currentOrigin == targetOrigin) {
+    init['headers'] = init['headers'] || {};
```
Comment (jridgewell) [inline src/utils/xhr-utils.js]: Nit: This should be [guaranteed](https://github.com/ampproject/amphtml/pull/17206/files#diff-6870dab961227bda821a287600fb7158R281).

File: src/utils/xhr-utils.js (hunk 10 above)
Comment (prateekbh) [reply src/utils/xhr-utils.js]: everything below this line should be exactly same.

File: src/utils/xhr-utils.js [hunk 15]
```diff
@@ -0,0 +1,626 @@
...
+    init['headers']['AMP-Same-Origin'] = 'true';
+  }
+  // In edge a `TypeMismatchError` is thrown when body is set to null.
+  dev().assert(init.body !== null, 'fetch `body` can not be `null`');
```
Comment (jridgewell) [inline src/utils/xhr-utils.js]: Nit: seems like we could move this into `setupInit`.

File: src/utils/xhr-utils.js (hunk 7 above)
Comment (jridgewell) [reply src/utils/xhr-utils.js]: Approved.

File: src/utils/xhr-utils.js (hunk 9 above)
Comment (jridgewell) [reply src/utils/xhr-utils.js]: Approved.

File: src/document-fetcher.js (hunk 13 above)
Comment (prateekbh) [reply src/document-fetcher.js]: sure

**Pull request merged**